# Ejercicios 2 de Pandas

---
## Ejercicio 1
Crea un DataFrame de 20x6 con los siguientes tipos de datos:
- Columnas numéricas con enteros y flotantes.
- Columna de fechas aleatorias en 2025.
- Columna categórica con valores como 'A', 'B', 'C'.
- Columna booleana aleatoria.
Muestra las primeras 5 filas y la estructura del DataFrame (`info()`).

In [ ]:
import pandas as pd
import numpy as np

col_int = np.random.randint(0, 100, 20)
# Multiplico por 100 para obtener un rango más amplio y no solo entre 0 y 1
col_float = np.random.random(20) * 100
# Fechas aleatorias en 2025
fechas = pd.to_datetime(
    np.random.randint(
        pd.Timestamp('2025-01-01').value // 10**9,
        pd.Timestamp('2025-12-31').value // 10**9,
        20
    ),
    unit='s'
)

# Choice elige aleatoriamente entre las opciones dadas
categorias = np.random.choice(['A', 'B', 'C'], size=20)
booleanos = np.random.choice([True, False], size=20)

df = pd.DataFrame({
    'Entero': col_int,
    'Float': col_float,
    'Fecha': fechas,
    'Categoría': categorias,
    'Booleano': booleanos
})

print("Primeras 5 filas:")
print(df.head(), "\n")
print("Información del DataFrame:")
print(df.info())



Primeras 5 filas:
   Entero      Float               Fecha Categoría  Booleano
0      26  18.635577 2025-09-14 14:43:27         A      True
1      23  46.282867 2025-01-27 08:44:10         B     False
2       6  86.061876 2025-07-15 10:28:03         B      True
3      51  58.060292 2025-01-11 13:40:01         C     False
4      76  41.062248 2025-11-27 02:22:37         A      True 

Información del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Entero     20 non-null     int32         
 1   Float      20 non-null     float64       
 2   Fecha      20 non-null     datetime64[ns]
 3   Categoría  20 non-null     object        
 4   Booleano   20 non-null     bool          
dtypes: bool(1), datetime64[ns](1), float64(1), int32(1), object(1)
memory usage: 708.0+ bytes
None
[2]


---
## Ejercicio 2
Introduce valores nulos en aproximadamente el 15% de las celdas del DataFrame.
- Reemplaza los nulos en columnas numéricas con la **media**.
- Reemplaza los nulos en columnas categóricas con el **valor más frecuente**.
- Muestra el resultado y verifica que no quedan nulos.

---
## Ejercicio 3
Crea un **MultiIndex** a partir de dos columnas categóricas y reorganiza el DataFrame usando este MultiIndex.
- Muestra un resumen por grupo usando `groupby` y calcula la media de las columnas numéricas.

---
## Ejercicio 4
Crea un DataFrame de 100 filas con tres columnas numéricas.
- Calcula la **correlación condicional** entre las columnas solo para las filas donde la columna booleana sea True.
- Identifica la pareja de columnas con mayor correlación condicional.

---
## Ejercicio 5
Genera un DataFrame con ventas de 50 productos en 12 meses (columna 'Producto', columnas 'Mes_1' a 'Mes_12').
- Calcula la suma anual por producto.
- Crea una columna con la **media mensual** de cada producto.
- Ordena el DataFrame por total anual descendente.

---
## Ejercicio 6
Lee un archivo CSV con al menos 500 filas y múltiples columnas (simula datos reales).
- Muestra las 10 filas con los valores más altos de una columna específica.
- Filtra filas que cumplan múltiples condiciones combinadas.

---
## Ejercicio 7
Crea un DataFrame de 30 filas con columnas numéricas y categóricas.
- Usa `pivot_table` para obtener la media de una columna numérica por categoría.
- Incluye un `margins=True` para obtener totales.

---
## Ejercicio 8
Crea un DataFrame con 100 filas, 5 columnas numéricas y 2 columnas de texto.
- Aplica una función con `apply` que transforme los valores numéricos según una condición (por ejemplo, multiplicar por 2 si >50).
- Aplica `applymap` a todo el DataFrame numérico para normalizar los valores entre 0 y 1.

---
## Ejercicio 9
Simula un escenario de base de datos:
- Crea dos DataFrames (`clientes` y `pedidos`) con columnas de id, nombre, producto y cantidad.
- Realiza un `merge` usando `id_cliente` para combinar información.
- Calcula el total de pedidos por cliente.

---
## Ejercicio 10
Crea un DataFrame de 50 filas con columna de fechas aleatorias.
- Agrupa por **mes y día de la semana** y calcula la media de otra columna numérica.
- Representa la tabla resultante en un DataFrame pivotado para ver los valores organizados por mes y día.